File: src/pydrology/models/hbv_light/processes.py
"""HBV-light snow and soil moisture process functions.

Numba-compiled functions implementing the snow routine (precipitation partitioning,
accumulation, melt, refreezing) and soil moisture routine (recharge, evapotranspiration).
"""

# ruff: noqa: I001
# Import order matters: _compat must patch numpy before numba import
import pydrology._compat  # noqa: F401

from numba import njit


@njit(cache=True)
def partition_precipitation(precip: float, temp: float, tt: float, sfcf: float) -> tuple[float, float]:
    """Partition precipitation into rain and snow.

    Args:
        precip: Total precipitation [mm/day].
        temp: Air temperature [deg C].
        tt: Threshold temperature [deg C].
        sfcf: Snowfall correction factor [-].

    Returns:
        Tuple of (p_rain, p_snow) in mm/day.
    """
    if temp > tt:
        return precip, 0.0
    else:
        return 0.0, sfcf * precip


@njit(cache=True)
def compute_melt(temp: float, tt: float, cfmax: float, snow_pack: float) -> float:
    """Compute snowmelt using degree-day method.

    Args:
        temp: Air temperature [deg C].
        tt: Threshold temperature [deg C].
        cfmax: Degree-day factor [mm/deg C/day].
        snow_pack: Current snow pack [mm].

    Returns:
        Melt rate [mm/day], limited by available snow.
    """
    if temp > tt:
        melt = cfmax * (temp - tt)
        return min(melt, snow_pack)
    return 0.0


@njit(cache=True)
def compute_refreezing(temp: float, tt: float, cfmax: float, cfr: float, liquid_water: float) -> float:
    """Compute refreezing of liquid water in snowpack.

    Args:
        temp: Air temperature [deg C].
        tt: Threshold temperature [deg C].
        cfmax: Degree-day factor [mm/deg C/day].
        cfr: Refreezing coefficient [-].
        liquid_water: Liquid water in snowpack [mm].

    Returns:
        Refreezing rate [mm/day], limited by available liquid water.
    """
    if temp < tt:
        refreeze = cfr * cfmax * (tt - temp)
        return min(refreeze, liquid_water)
    return 0.0


@njit(cache=True)
def update_snow_pack(
    sp: float,
    lw: float,
    p_snow: float,
    melt: float,
    refreeze: float,
    cwh: float,
) -> tuple[float, float, float]:
    """Update snow pack and liquid water states.

    Args:
        sp: Current snow pack [mm].
        lw: Current liquid water in snow [mm].
        p_snow: Snowfall [mm/day].
        melt: Snowmelt [mm/day].
        refreeze: Refreezing [mm/day].
        cwh: Water holding capacity of snow [-].

    Returns:
        Tuple of (new_sp, new_lw, outflow):
        - new_sp: Updated snow pack [mm]
        - new_lw: Updated liquid water in snow [mm]
        - outflow: Water released from snow to soil [mm/day]
    """
    # Update snow pack: add snowfall, subtract melt, add refrozen water
    new_sp = sp + p_snow - melt + refreeze

    # Update liquid water: add melt, subtract refreezing
    new_lw = lw + melt - refreeze

    # Compute maximum liquid water storage
    lw_max = cwh * new_sp

    # Compute outflow (excess liquid water)
    if new_lw > lw_max:
        outflow = new_lw - lw_max
        new_lw = lw_max
    else:
        outflow = 0.0

    # Ensure non-negative
    new_sp = max(0.0, new_sp)
    new_lw = max(0.0, new_lw)

    return new_sp, new_lw, outflow


@njit(cache=True)
def compute_recharge(soil_input: float, sm: float, fc: float, beta: float) -> float:
    """Compute groundwater recharge from soil.

    Uses the HBV soil moisture / runoff relationship:
    Recharge/Input = (SM/FC)^BETA

    Args:
        soil_input: Water input to soil (rain + snowmelt) [mm/day].
        sm: Current soil moisture [mm].
        fc: Field capacity [mm].
        beta: Shape coefficient [-].

    Returns:
        Recharge to groundwater [mm/day].
    """
    if fc <= 0.0 or soil_input <= 0.0:
        return 0.0

    # Saturation ratio, clamped to [0, 1]
    sr = min(max(sm / fc, 0.0), 1.0)

    # Recharge fraction
    recharge = soil_input * (sr**beta)

    return recharge


@njit(cache=True)
def compute_actual_et(pet: float, sm: float, fc: float, lp: float) -> float:
    """Compute actual evapotranspiration from soil.

    ET is reduced linearly when SM < LP * FC.

    Args:
        pet: Potential evapotranspiration [mm/day].
        sm: Current soil moisture [mm].
        fc: Field capacity [mm].
        lp: Limit for potential ET as fraction of FC [-].

    Returns:
        Actual evapotranspiration [mm/day].
    """
    if fc <= 0.0 or lp <= 0.0:
        return 0.0

    lp_threshold = lp * fc

    # Linear reduction below threshold, full PET above
    et_act = pet if sm >= lp_threshold else pet * sm / lp_threshold

    # Cannot extract more than available
    et_act = min(et_act, max(sm, 0.0))

    return et_act


@njit(cache=True)
def update_soil_moisture(
    sm: float,
    soil_input: float,
    recharge: float,
    et_act: float,
    fc: float,
) -> float:
    """Update soil moisture state.

    Args:
        sm: Current soil moisture [mm].
        soil_input: Water input to soil [mm/day].
        recharge: Recharge leaving to groundwater [mm/day].
        et_act: Actual evapotranspiration [mm/day].
        fc: Field capacity [mm].

    Returns:
        Updated soil moisture [mm], bounded [0, FC].
    """
    # Infiltration is the non-recharge portion
    infiltration = soil_input - recharge

    # Update soil moisture
    new_sm = sm + infiltration - et_act

    # Bound to [0, FC]
    new_sm = max(0.0, min(new_sm, fc))

    return new_sm

File: crates/pydrology-core/src/hbv_light/processes.rs
//! HBV-Light snow and soil moisture process functions.
//!
//! Pure functions implementing the snow routine (precipitation partitioning,
//! accumulation, melt, refreezing) and soil moisture routine (recharge, ET).

/// Partition precipitation into rain and snow.
///
/// Returns (p_rain, p_snow).
pub fn partition_precipitation(precip: f64, temp: f64, tt: f64, sfcf: f64) -> (f64, f64) {
    if temp > tt {
        (precip, 0.0)
    } else {
        (0.0, sfcf * precip)
    }
}

/// Compute snowmelt using degree-day method.
///
/// Melt is limited by available snow pack.
pub fn compute_melt(temp: f64, tt: f64, cfmax: f64, snow_pack: f64) -> f64 {
    if temp > tt {
        let melt = cfmax * (temp - tt);
        melt.min(snow_pack)
    } else {
        0.0
    }
}

/// Compute refreezing of liquid water in snowpack.
///
/// Refreezing is limited by available liquid water.
pub fn compute_refreezing(temp: f64, tt: f64, cfmax: f64, cfr: f64, liquid_water: f64) -> f64 {
    if temp < tt {
        let refreeze = cfr * cfmax * (tt - temp);
        refreeze.min(liquid_water)
    } else {
        0.0
    }
}

/// Update snow pack and liquid water states.
///
/// Returns (new_sp, new_lw, outflow).
pub fn update_snow_pack(
    sp: f64,
    lw: f64,
    p_snow: f64,
    melt: f64,
    refreeze: f64,
    cwh: f64,
) -> (f64, f64, f64) {
    let mut new_sp = sp + p_snow - melt + refreeze;
    let mut new_lw = lw + melt - refreeze;

    let lw_max = cwh * new_sp;

    let outflow = if new_lw > lw_max {
        let excess = new_lw - lw_max;
        new_lw = lw_max;
        excess
    } else {
        0.0
    };

    new_sp = new_sp.max(0.0);
    new_lw = new_lw.max(0.0);

    (new_sp, new_lw, outflow)
}

/// Compute groundwater recharge from soil.
///
/// Uses the HBV soil moisture / runoff relationship: Recharge/Input = (SM/FC)^BETA.
pub fn compute_recharge(soil_input: f64, sm: f64, fc: f64, beta: f64) -> f64 {
    if fc <= 0.0 || soil_input <= 0.0 {
        return 0.0;
    }
    let sr = (sm / fc).clamp(0.0, 1.0);
    soil_input * sr.powf(beta)
}

/// Compute actual evapotranspiration from soil.
///
/// ET is reduced linearly when SM < LP * FC.
pub fn compute_actual_et(pet: f64, sm: f64, fc: f64, lp: f64) -> f64 {
    if fc <= 0.0 || lp <= 0.0 {
        return 0.0;
    }

    let lp_threshold = lp * fc;

    let et_act = if sm >= lp_threshold {
        pet
    } else {
        pet * sm / lp_threshold
    };

    et_act.min(sm.max(0.0))
}

/// Update soil moisture state.
///
/// Returns updated soil moisture, bounded to [0, FC].
pub fn update_soil_moisture(
    sm: f64,
    soil_input: f64,
    recharge: f64,
    et_act: f64,
    fc: f64,
) -> f64 {
    let infiltration = soil_input - recharge;
    let new_sm = sm + infiltration - et_act;
    new_sm.clamp(0.0, fc)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn assert_approx(actual: f64, expected: f64, tol: f64) {
        assert!(
            (actual - expected).abs() < tol,
            "expected {expected} +/- {tol}, got {actual}"
        );
    }

    // -- Precipitation partitioning --

    #[test]
    fn rain_above_threshold() {
        let (p_rain, p_snow) = partition_precipitation(10.0, 5.0, 0.0, 1.0);
        assert_eq!(p_rain, 10.0);
        assert_eq!(p_snow, 0.0);
    }

    #[test]
    fn snow_below_threshold() {
        let (p_rain, p_snow) = partition_precipitation(10.0, -5.0, 0.0, 1.0);
        assert_eq!(p_rain, 0.0);
        assert_eq!(p_snow, 10.0);
    }

    #[test]
    fn sfcf_correction() {
        let (p_rain, p_snow) = partition_precipitation(10.0, -5.0, 0.0, 0.8);
        assert_eq!(p_rain, 0.0);
        assert_eq!(p_snow, 8.0);
    }

    // -- Snowmelt --

    #[test]
    fn melt_above_threshold() {
        let melt = compute_melt(5.0, 0.0, 3.0, 100.0);
        assert_approx(melt, 15.0, 1e-10);
    }

    #[test]
    fn no_melt_below_threshold() {
        let melt = compute_melt(-5.0, 0.0, 3.0, 100.0);
        assert_eq!(melt, 0.0);
    }

    #[test]
    fn melt_limited_by_snow() {
        let melt = compute_melt(10.0, 0.0, 3.0, 5.0);
        assert_eq!(melt, 5.0);
    }

    // -- Refreezing --

    #[test]
    fn refreeze_below_threshold() {
        let refreeze = compute_refreezing(-5.0, 0.0, 3.0, 0.05, 10.0);
        assert_approx(refreeze, 0.75, 1e-10);
    }

    #[test]
    fn no_refreeze_above_threshold() {
        let refreeze = compute_refreezing(5.0, 0.0, 3.0, 0.05, 10.0);
        assert_eq!(refreeze, 0.0);
    }

    #[test]
    fn refreeze_limited_by_liquid() {
        let refreeze = compute_refreezing(-10.0, 0.0, 3.0, 0.5, 2.0);
        assert_eq!(refreeze, 2.0);
    }

    // -- Snow pack update --

    #[test]
    fn snow_pack_outflow_when_lw_exceeds_capacity() {
        let (new_sp, new_lw, outflow) = update_snow_pack(10.0, 0.0, 0.0, 5.0, 0.0, 0.1);
        // new_sp = 10 + 0 - 5 + 0 = 5
        // new_lw = 0 + 5 - 0 = 5
        // lw_max = 0.1 * 5 = 0.5
        // outflow = 5 - 0.5 = 4.5
        assert_approx(new_sp, 5.0, 1e-10);
        assert_approx(new_lw, 0.5, 1e-10);
        assert_approx(outflow, 4.5, 1e-10);
    }

    // -- Recharge --

    #[test]
    fn recharge_increases_with_moisture() {
        let low_sm = compute_recharge(10.0, 50.0, 250.0, 2.0);
        let high_sm = compute_recharge(10.0, 200.0, 250.0, 2.0);
        assert!(high_sm > low_sm);
    }

    #[test]
    fn recharge_at_saturation() {
        let recharge = compute_recharge(10.0, 250.0, 250.0, 2.0);
        assert_approx(recharge, 10.0, 1e-10);
    }

    #[test]
    fn no_recharge_when_no_input() {
        let recharge = compute_recharge(0.0, 125.0, 250.0, 2.0);
        assert_eq!(recharge, 0.0);
    }

    // -- Actual ET --

    #[test]
    fn full_et_above_lp_threshold() {
        let et = compute_actual_et(5.0, 225.0, 250.0, 0.9);
        assert_approx(et, 5.0, 1e-10);
    }

    #[test]
    fn reduced_et_below_lp_threshold() {
        let et = compute_actual_et(5.0, 112.5, 250.0, 0.9);
        assert_approx(et, 2.5, 1e-10);
    }

    #[test]
    fn et_limited_by_available() {
        let et = compute_actual_et(300.0, 230.0, 250.0, 0.9);
        assert_approx(et, 230.0, 1e-10);
    }

    // -- Soil moisture update --

    #[test]
    fn soil_moisture_bounded_by_fc() {
        let new_sm = update_soil_moisture(240.0, 100.0, 10.0, 5.0, 250.0);
        assert!(new_sm <= 250.0);
        assert!(new_sm >= 0.0);
    }

    #[test]
    fn soil_moisture_non_negative() {
        let new_sm = update_soil_moisture(5.0, 0.0, 0.0, 100.0, 250.0);
        assert_eq!(new_sm, 0.0);
    }
}
